crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+
+Because a sealed type is equivalent to the union of all of its proper subtypes,
+for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`. To give a Python example:
+
+```py
+from enum import Enum, auto
+
+class X(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)
+
+    if var is X.A:
+        ...  # var is narrowed to `Literal[X.A]`
+    else:
+        ...  # var is narrowed to `Literal[X.B, X.C]`
+```
+
+## Single-value types
+
+A single-value type is a non-empty type for which it is known
+that all inhabitants of the type are equivalent with respect to their runtime value.
+All singleton types are single-value types, but not all single-value types are singleton types.
+
+Examples of single-value types that are not singleton types
+are `Literal["foo"]`, `Literal[b"foo"]`, `Literal[123456]`, and `tuple[Literal[True], Literal[False]]`.
+All runtime inhabitants of the `Literal["foo"]` type are entirely fungible and equal
+(they are all equal to the string `"foo"`).
+However, they are not necessarily the *same* object in terms of identity;
+multiple instances of the `"foo"` string can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
+
+In order for a type to be considered a single-value type,
+there must exist a reflexive, symmetric, and transitive equivalence relation
+between all inhabitants of the type. It follows that for a single-value type in Python,
+all inhabitants must be instances of the same runtime class,
+and the class in question must have `__eq__` and `__ne__` methods
+such that the equality relation between instances of the class
+is reflexive, symmetric, and transitive.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+## Closed types
+
+A closed type is a type for which it is known that the type has no subtypes
+except for the type itself and `Never`; it is "closed for extension".
+For a given closed type `X`, the only subtypes of `X` are `X` and `Never`;
+the only *proper* subtype of `X` is `Never`.
```

Comment:
I think this wording has a similar problem with arbitrary intersection and union types as the above wording on sealed types.

I think we have to present closed types in a way that's more tied to "instance" types -- as in, it's a type defined as instances of a class, which cannot be subclassed, therefore all inhabitants of the type must be instances of the final class.

Reply:
> I think we have to present closed types in a way that's more tied to "instance" types -- as in, it's a type defined as instances of a class, which cannot be subclassed, therefore all inhabitants of the type must be instances of the final class.

I'm not sure we want the concept of "closed" types to be limited to _just_ `@final` nominal types, though. I think we at least want `TypedDict`s with `closed=True` to qualify (assuming [PEP 728](https://peps.python.org/pep-0728/) is accepted)?